addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/check-launch-dependency/check-launch-dependency.js

```diff
@@ -82,8 +80,25 @@ class CheckLaunchDependency extends StepBase {
       this.payloadOrConfig.string(inPayloadKeys.envTypeId),
       this.payloadOrConfig.string(inPayloadKeys.envTypeConfigId),
     ]);
+    const [albService, lockService, envTypeConfigService] = await this.mustFindServices([
+      'albService',
+      'lockService',
+      'envTypeConfigService',
+    ]);
     const projectId = resolvedVars.projectId;
-    const [envTypeConfigService] = await this.mustFindServices(['envTypeConfigService']);
+    const awsAccountId = await albService.findAwsAccountId(requestContext, projectId);
+    // Locking the ALB provisioing to avoid race condiitons on parallel provisioning.
+    // expiresIn is set to 10 minutes. attemptsCount is set to 600 to retry after 1 seconds for 10 minutes
+    const lock = await lockService.tryWriteLock(
+      { id: `alb-update-${awsAccountId}`, expiresIn: 600 },
+      { attemptsCount: 600 },
+    );
+    this.print({
+      msg: `obtained lock - ${lock}`,
+    });
+    if (_.isUndefined(lock)) throw new Error('Could not obtain a lock');
+    this.state.setKey('ALB_LOCK', lock);
```

Comment:
This seems brittle to me. What if your lock in DDB actually expires(say after 10 minutes) but the variable here isn't reset because step function wasn't done yet? In that case the step function would assume that it has the lock when it doesn't really have it. Can't we add lock on the caller of step function instead so that it is performed cleanly under a lock?

Reply:
Added additional checks to see if lock is present as per discussion